oldps.1
.\" This file Copyright 1992, 1997 Michael K. Johnson <johnsonm@redhat.com>
.\" and 1996 Charles L. Blake <cblake@bbn.com>
.\" It may be distributed under the GNU Public License, version 2, or
.\" any higher version.  See section COPYING of the GNU Public license
.\" for conditions under which this file may be redistributed.
.TH PS 1 "3 Sep 1997" "Cohesive Systems" "Linux User's Manual"
.SH NAME
ps \- report process status
.SH SYNOPSIS
ps [\fBlujsvmaxScewhrnu\fR] [\fBt\fIxx\fR] \
[\fBO\fR[\fB+\fR|\fB-\fR]\fIk1\fR[[\fB+\fR|\fB-\fR]\fIk2\fR...]] \
[\fIpids\fR]

there are also three long options:

.BR \-\-sort\fIX [ + | - ] \fIkey [,[ + | - ] \fIkey [, ... ]]

.B "\-\-help"

.B "\-\-version"

More long options are on the way...
.SH DESCRIPTION
.B "ps "
gives a snapshot of the current processes.  If you want a repetitive
update of this status, use
.BR top .
This man page documents the
.IR /proc -based
version of
.BR ps ,
or tries to.
.PP
.SH "COMMAND-LINE OPTIONS"
The command-line options for this version of
.B ps
are derived from the BSD version of
.BR ps ,
not the System V version.

The command-line arguments should \fBnot\fP be preceeded by a `\-' character,
because in the future, a `\-' will be used to indicate Unix98-standard
command-line arguments, while no `\-' will indicate the current
``extended BSD'' style of command line arguments.

For now, ps will give you a warning if you use a `\-' for a short option,
but it will still work.  If you have shell scripts which use BSD-style
arguments to ps, take heed of the warning and fix them, or else your
scripts will fail to function correctly at some point in the future.
If you want to turn off the warnings, set the
.B I_WANT_A_BROKEN_PS
environment variable. 

There are also some ``long options'' in GNU style; see below for those.
.PP
.PD 0
.TP 0.5i
.B "l "
long format
.TP 0.5i
.B "u "
user format: gives user name and start time
.TP 0.5i
.B "j "
jobs format: pgid sid
.TP 0.5i
.B "s"
signal format
.TP 0.5i
.B "v "
vm format
.TP 0.5i
.B "m "
displays memory info (combine with
.B p
flag to get number of pages).
.TP 0.5i
.B "f "
"forest" family tree format for command line
.TP 0.5i
.B "a "
show processes of other users too
.TP 0.5i
.B "x "
show processes without controlling terminal
.TP 0.5i
.B "S "
add child cpu time and page faults
.TP 0.5i
.B "c "
command name from task_struct
.TP 0.5i
.B "e "
show environment after command line and ` + '
.TP 0.5i
.B "w "
wide output: don't truncate command lines to fit on one line.
To be exact, every w that is specified will add another possible
line to the output. If the space isn't needed it isn't used. You
may up to 100
.BR w 's.
.TP 0.5i
.B "h "
no header
.TP 0.5i
.B "r "
running procs only
.TP 0.5i
.B "n "
numeric output for
.BR USER " and " WCHAN .
.PD 1
.TP 0.5i
.BI t xx
only procs with controlling tty \fIxx\fR; for \fIxx\fR you may use either the
name of a device file under "/dev" or that name with either
.IR tty " or " cu
sliced off.  This is the reverse heuristic that ps uses to print out the
abbreviated tty name in the \fBTT\fR field, e.g.
.BR "ps t1" .
.TP 0.5i
.BR O [ + | - ] \fIk1 [,[ + | - ] \fIk2 [, ... ]]
Order the process listing according to the multi-level sort specified by
the sequence of \fIshort\fR keys from \fBSORT KEYS\fR, \fIk1\fR, \fIk2\fR, ...
Default order specifications exist for each of the various formats of \fBps\fR.
These are over-ridden by a user specified ordering.  The `+' is quite optional,
merely re-iterating the default direction on a key.  `-' reverses direction only
on the key it precedes.  As with \fBt\fR and \fIpids\fR, the O option must be
the last option in a single command argument, but specifications in successive
arguments are catenated.
.TP 0.5i
.I pids
List only the specified processes; they are comma-delimited.  The 
list must be given immediately after the last option in a single command-line
argument, with no intervening space, e.g.
.BR "ps j1,4,5" .
Lists specified in subsequent arguments are catenated, e.g.
.B ps l 1,2 3,4 5 6
will list all of the processes 1-6 in long format.  If pids are given, they
are listed no matter what.  If a tty is given matching processes are listed
no matter what.  These two features override the 'a' and 'x' flags.
.SH "LONG COMMAND\-LINE OPTIONS"
These options are preceeded by a double\-hyphen.
.TP 0.5i
.BR \-\-sort\fIX [ + | - ] \fIkey [,[ + | - ] \fIkey [, ... ]]
Choose a \fImulti-letter key\fR from the \fBSORT KEYS\fR section. \fIX\fR may be
any convenient separator character.  To be GNU-ish use `='.  The `+' is really
optional since default direction is increasing numerical or lexicographic order.
E.g.:
.B ps jax --sort=uid,-ppid,+pid
.TP 0.5i
.B "\-\-help"
Get a help message that summarizes the usage and gives a list of
supported sort keys.  This list may be more up to date than this man
page.
.TP 0.5i
.B "\-\-version"
Display version and source of this program.
.SH "SORT KEYS"
Note that the values used in sorting are the internal values \fBps\fR uses and
\fInot\fR the `cooked' values used in some of the output format fields.  If
someone wants to volunteer to write special comparison functions for the cooked
values, ... ;-)

SHORT   LONG            DESCRIPTION
.PD 0
.TP 0.5i
c       cmd             simple name of executable
.TP 0.5i
C       cmdline         full command line
.TP 0.5i
f       flags           flags as in long format F field
.TP 0.5i
g       pgrp            process group ID
.TP 0.5i
G       tpgid           controlling tty process group ID
.TP 0.5i
j       cutime          cumulative user time
.TP 0.5i
J       cstime          cumulative system time
.TP 0.5i
k       utime           user time
.TP 0.5i
K       stime           system time
.TP 0.5i
m       min_flt         number of minor page faults
.TP 0.5i
M       maj_flt         number of major page faults
.TP 0.5i
n       cmin_flt        cumulative minor page faults
.TP 0.5i
N       cmaj_flt        cumulative major page faults
.TP 0.5i
o       session         session ID
.TP 0.5i
p       pid             process ID
.TP 0.5i
P       ppid            parent process ID
.TP 0.5i
r       rss             resident set size
.TP 0.5i
R       resident        resident pages
.TP 0.5i
s       size            memory size in kilobytes
.TP 0.5i
S       share           amount of shared pages
.TP 0.5i
t       tty             the minor device number of tty
.TP 0.5i
T       start_time      time process was started
.TP 0.5i
U       uid             user ID number
.TP 0.5i
u       user            user name
.TP 0.5i
v       vsize           total VM size in bytes
.TP 0.5i
y       priority        kernel scheduling priority
.PD 1
.SH "FIELD DESCRIPTIONS"
.TP 0.5i
.B "PRI "
This is the counter field in the task struct.  It is the time in
.B HZ
of the process's possible timeslice.
.TP 0.5i
.B "NI "
Standard unix nice value; a positive value means less cpu time.
.TP 0.5i
.B "SIZE "
Virtual image size; size of text+data+stack.
.TP 0.5i
.B "RSS "
Resident set size; kilobytes of program in memory.
.TP 0.5i
.B "WCHAN "
Name of the kernel function where the process is sleeping, with the
.RB ` sys_ '
stripped from the function name.  If
.I /etc/psdatabase
does not exist, it is just a hex number instead.
.TP 0.5i
.B "STAT "
Information about the status of the process.  The first field is
.B R
for runnable,
.B S
for sleeping,
.B D
for uninterruptible sleep,
.B T
for stopped or traced, or
.B Z
for a zombie process.  The second field contains
.B W
if the process has no resident pages.  The third field is
.B N
if the process has a positive nice value
.RB ( NI
field).
.TP 0.5i
.B "TT "
Controlling tty.
.TP 0.5i
.B "PAGEIN "

NEWS
procps-ng-NEXT
---------------
  * library
    internal: Various fixes for Hurd                       merge !303
  * build: add libpthread if needed                        issue #434
  * kill: Correct exit value for pgrp kills                issue #438
  * pkill: echo with other options                         issue #428
  * sysctl: Really ignore stable_secret                    issue #308

procps-ng-4.0.7
---------------
  * library
    version: inc revision to 2 now 1:2:0
    internal: procps_pids_length off by one                issue #412
    external: fix slabinfo header extern 'C' declaration   issue #415
    internal: fix file descriptor leaks in <pids> api      issue #421
    internal: strv items are now escaped in <pids> api     issue #429
    internal: fix output if on seconds edge values         merge !246 RHEL-60825
  * pidof: Add -d aliased option                           issue #418
  * pgrep: Don't treat empty list as 0                     issue #427
  * pmap: Fix testsuite for Alpha                          Debian #1141465
  * ps: correct 'environ' output when file unavailable
  * ps: minimize potential EACCES with 'environ' files     issue #431
  * top: avoid batch mode segfault with maximum width      issue #422
  * w: Correctly check for end of tty using utmp           issue #430
  * watch: Dont remove 2 lines when using -t option        issue #413
  * watch: Handle resizing better                          issue #417
  * watch: Restore LINES and COLUMNS env variables         issue #432
procps-ng-4.0.6
---------------
  * library
    version: inc revision to 1 now 1:1:0
    internal: Use openat for files under /proc/<PID>
    internal: Don't check for sd_booted                    Debian #1108549
    internal: Address potential race leading to segfault   issues #380, #390
  * local: guard SIGPOLL for MacOS                         merge !246
  * pgrep: Fix unhex breakage for signal numbers           issue #369
  * pgrep: Match on process ID                             Debian #612146
  * pgrep: Add process_mrelease                            merge !250
  * pgrep: Add shell quoting                               merge !248
  * pgrep: Use pidfd_send_signal()                         issue #207
  * pgrep.1: Note that we can only go to 100th of second   issue #376
  * pgrep: Add --quiet option                              issue #404
  * pmap: add -k option to print raw names from kernel     merge !251
  * ps: parse --help correctly                             issue #381
  * ps: Add --delimiter option                             issue #118
  * ps: ps f forest works under pid 1                      issue #102
  * ps: no longer fails with many pids & the 'p' option    issue #387
  * ps.1: cols and collums alias width option              Debian #926361
  * ps.1: Add format equivalents                           Debian #925437
  * ps: avoid potential segfault due to library race       issue #380
  * ps: Fix -ad option conflict                            issue #395
  * ps: ppid of 0 is ok, fix parser                        issue #405
  * slabtop: Increase column width                         Debian #959375
  * sysctl: Return error if EPERM or EISDIR                merge !228
  * sysctl: Use options after --system                     Debian #978989
  * top: provides for ignoring configuration file(s)       issue #365
  * top: fix legacy configuration file vulnerability       issue #384
  * top: added a new help screen for specialized keys
  * top: avoid potential segfault due to library race      issue #390
  * top: new feature to show physical core percentages
  * w: Add terminal mode to show all terminal sessions     issue #375
  * w: Use process TTY as backup for user TTY              Debian #1080335
  * w: Use correct return value for sd_get_sessions        Debian #1068904

oldps.c
/*
 * ps.c                - show process status
 *
 * Copyright (c) 1992 Branko Lankester
 *
 * Snarfed and HEAVILY modified for procps by Michael K. Johnson,
 * (johnsonm@sunsite.unc.edu).  What is used is what is required to have a
 *  common interface.
 *
 * Massive modifications by Charles Blake (cblake@bbn.com).  Rewrite
 * of the system process table code, multilevel sorting, device number
 * database, forest feature (contributed by ...), environment variables, GNU
 * style long options, pid list filtering (contributed by Michael Shields).
 *
 * Michael K. Johnson <johnsonm@redhat.com> again in charge, merging
 * patches that have accumulated over a long time.
 *
 * Changes Copyright (C) 1993, 1994, 1997 Michael K. Johnson,
 *   and   Copyright (C) 1995, 1996 Charles Blake
 * See file COPYING for copyright details.
 */
#include <stdio.h>
#include <stdlib.h>
#include <ctype.h>
#include <string.h>
#include <unistd.h>
#include <fcntl.h>
#include <pwd.h>
#include <time.h>
#include <limits.h>
#include <sys/types.h>
#include <sys/ioctl.h>
#include <sys/resource.h>
#include <sys/param.h>

#include <proc/version.h>
#include <proc/readproc.h>
#include <proc/procps.h>
#include <proc/devname.h>
#include <proc/tree.h>
#include <proc/sysinfo.h>
#include <proc/status.h>
#include <proc/compare.h>

static void show_short(char*, proc_t*);
static void show_long (char*, proc_t*);
static void show_user (char*, proc_t*);
static void show_jobs (char*, proc_t*);
static void show_sig  (char*, proc_t*);
static void show_vm   (char*, proc_t*);
static void show_m    (char*, proc_t*);
static void show_regs (char*, proc_t*);

/* this struct replaces the previously parallel fmt_fnc and hdrs */
static struct {
    void      (*format)(char*,proc_t*);
    char        CL_option_char;
    const char*	default_sort;
    const char*	header;
} mode[] = {
    { show_short,  0 , "Up", "  PID TTY STAT TIME COMMAND" },
    { show_long,  'l', "Pp", " FLAGS   UID   PID  PPID PRI  NI   SIZE   RSS WCHAN       STA TTY TIME COMMAND" },
    { show_user,  'u', "up", "USER       PID %CPU %MEM  SIZE   RSS TTY STAT START   TIME COMMAND" },
    { show_jobs,  'j', "gPp"," PPID   PID  PGID   SID TTY TPGID  STAT  UID   TIME COMMAND" },
    { show_sig,   's', "p",  "  UID   PID SIGNAL   BLOCKED  IGNORED  CATCHED  STAT TTY   TIME COMMAND" },
    { show_vm,    'v', "r",  "  PID TTY STAT TIME  PAGEIN TSIZ DSIZ  RSS   LIM %MEM COMMAND" },
    { show_m,     'm', "r",  "  PID TTY MAJFLT MINFLT   TRS   DRS  SIZE  SWAP   RSS  SHRD   LIB  DT COMMAND" },
    { show_regs,  'X', "p",  "NR   PID    STACK      ESP      EIP TMOUT ALARM STAT TTY   TIME COMMAND" },
    { NULL, 0, NULL, NULL }
};

/* some convenient integer constants corresponding to the above rows. */
enum PS_MODALITY { PS_D = 0, PS_L, PS_U, PS_J, PS_S, PS_V, PS_M, PS_X };

static char * prtime         (char *s, unsigned long t, unsigned long rel);
static void   usage          (char* context);
static void   set_cmdspc     (int w_opts);
static void   show_procs     (unsigned maxcmd, int do_header, int, void*,int);
static void   show_cmd_env   (char* tskcmd, char** cmd, char** env, unsigned maxch);
static void   show_a_proc    (proc_t* the_proc, unsigned maxcmd);
static void   show_time      (char *s, proc_t * p);
static void   add_node       (char *s, proc_t *task);
static int    node_cmp       (const void *s1, const void *s2);
static void   show_tree      (int n, int depth, char *continued);
static void   show_forest    (void);

static int CL_fmt       = 0;

/* process list filtering command line options */

static int CL_all,
    CL_kern_comm,
    CL_no_ctty,
    CL_run_only;
static char  * CL_ctty;
static pid_t * CL_pids;          /* zero-terminated list, dynamically allocated */

/* process display modification command line options */

static int CL_show_env,
    CL_num_outp,          /* numeric fields for user, wchan, tty */
    CL_sort     = 1,
    CL_forest,
    CL_Sum,
    CL_pg_shift = (PAGE_SHIFT - 10);	/* default: show k instead of pages */

/* Globals */

static unsigned cmdspc = 80;     /* space left for cmd+env after table per row */
static int      GL_current_time; /* some global system parameters */
static long     GL_time_now;
static int      GL_wchan_nout;   /* this is can also be set on the command-line  */


/*****************************/
int main(int argc, char **argv) {
    char *p;
    int width = 0,
	do_header = 1,
	user_ord = 0,
	next_arg = 0,
	toppid = 0,
	pflags, N = 1;
    void* args = NULL;
    dev_t tty[2] = { 0 };
    uid_t uid[1];
    
    do {
        --argc;		/* shift to next arg. */
        ++argv;
        for (p = *argv; p && *p; ++p) {
            switch (*p) {
	      case '-':               /* "--" ==> long name options */
		if (*(p+1) == '-') {
                    if (strncmp(p+2,"sort",4)==0) {
			if (parse_long_sort(p+7))
			    usage("unrecognized long sort option\n");
			user_ord = 1;
			next_arg = 1;
			break;
                    } else if (strncmp(p+2, "help", 4) == 0) {
			usage(NULL);
			return 0;
                    } else if (strncmp(p+2, "version", 6) == 0) {
			display_version();
			return 0;
                    } else if (*(p+2) != '\0')	/* just '-- '; not an error */
			usage("ps: unknown long option\n");
		}
		if (!getenv("I_WANT_A_BROKEN_PS")) {
		    fprintf(stderr,
			    "warning: `-' deprecated; use `ps %s', not `ps %s'\n",
			    (p+1), p);
		}
		break;
	      case 'l': CL_fmt = PS_L;   /* replaceable by a */	break;
	      case 'u': CL_fmt = PS_U;   /* loop over mode[] */	break;
	      case 'j': CL_fmt = PS_J;				break;
	      case 's': CL_fmt = PS_S;				break;
	      case 'v': CL_fmt = PS_V;				break;
	      case 'm': CL_fmt = PS_M;				break;
	      case 'X': CL_fmt = PS_X;   /* regs */		break;

	      case 'r': CL_run_only = 1; /* list filters */	break;
	      case 'a': CL_all = 1;				break;
	      case 'x': CL_no_ctty = 1;				break;
	      case 't': CL_ctty = p + 1;
		next_arg = 1;				break;

	      case 'e': CL_show_env = 1; /* output modifiers */	break;
	      case 'f': CL_forest = 1;
		CL_kern_comm = 0;			break;
	      case 'c': CL_kern_comm = 1;			break;
	      case 'w': ++width;				break;
	      case 'h': do_header = 0;				break;
	      case 'n': CL_num_outp = 1;
		GL_wchan_nout = 1;			break;
	      case 'S': CL_Sum = 1;				break;
	      case 'p': CL_pg_shift = 0;			break;
	      case 'o': CL_sort = !CL_sort;			break;
	      case 'O':
		if (parse_sort_opt(p+1))
		    usage("short form sort flag parse error\n");
		user_ord = 1;
		next_arg = 1;
		break;
	      case 'V': display_version(); exit(0);
	      default:
                /* Step through, reading+alloc space for comma-delim pids */
		if (isdigit(*p)) {
		    while (isdigit(*p)) {
			CL_pids = xrealloc(CL_pids, (toppid + 2)*sizeof(pid_t));
			CL_pids[toppid++] = atoi(p);
			while (isdigit(*p))
			    p++;
			if (*p == ',')
			    p++;
		    }
		    CL_pids[toppid] = 0;
		    next_arg = 1;
		}
		if (*p)
		    usage("unrecognized option or trailing garbage\n");
            }
            if (next_arg) {
                next_arg = 0;
                break;       /* end loop over chars in this argument */
            }
        }
    } while (argc > 1);

    meminfo();

    if (!CL_sort)	/* since the unsorted mode is intended to be speedy */
	CL_forest = 0;	/* turn off the expensive forest option as well. */

    if (CL_fmt == PS_L) {
	if (open_psdb(NULL)) GL_wchan_nout = 1;
    }

    set_cmdspc(width);

    if (!(GL_current_time = uptime(0,0)))
	return 1;
    GL_time_now = time(0L);

    if (CL_sort && !user_ord)
        parse_sort_opt(mode[CL_fmt].default_sort);

    /* NOTE:  all but option parsing has really been done to enable
     * multiple uid/tty/state filtering as well as multiple pid filtering
     */
    pflags = PROC_ANYTTY;	/* defaults */

    if (!CL_kern_comm)	pflags |= PROC_FILLCMD;  	 /* verbosity flags */
    if (CL_fmt == PS_M) pflags |= PROC_FILLMEM;
    if (CL_show_env)	pflags |= PROC_FILLENV;
    if (!CL_num_outp)	pflags |= PROC_FILLUSR;

    if (CL_no_ctty)	pflags &= ~PROC_ANYTTY;		/* filter flags */
    if (CL_run_only)  { pflags |= PROC_STAT; args = "RD"; }
    else if (!CL_all)      { pflags |= PROC_UID;  args = uid; uid[0] = getuid(); pflags &= ~PROC_STAT; }
    if (CL_pids)      { pflags |= PROC_PID;  args = CL_pids; pflags &= ~PROC_UID; pflags &= ~PROC_STAT; }
    if (CL_ctty) {
	if ((tty[0] = tty_to_dev(CL_ctty)) == (dev_t)-1) {
	    fprintf(stderr, "the name `%s' is not a tty\n", CL_ctty);
	    exit(1);
	}
	pflags = (pflags | PROC_TTY) & ~(PROC_ANYTTY|PROC_STAT|PROC_UID|PROC_PID);
	args = tty;
    }
    show_procs(cmdspc, do_header, pflags, args, N);
    return 0;
}

/***************************/
/* print a context dependent usage message and maybe exit */
static void usage(char* context) {
    fprintf(stderr,
	    "%s"
            "usage:  ps acehjlnrsSuvwx{t<tty>|#|O[-]u[-]U..} \\\n"
            "           --sort:[-]key1,[-]key2,...\n"
            "           --help gives you this message\n"
            "           --version prints version information\n",
	    context ? context : "");
    if (context)
	exit(1);	/* prevent bad exit status by calling usage(NULL) */
}

/*****************************/
/* set maximum chars displayed on a line based on screen size.
 * Always allow for the header, with n+1 lines of output per row.
 */
static void set_cmdspc(int n) {
    struct winsize win;
    int h = strlen(mode[CL_fmt].header),
	c = strlen("COMMAND");

    if (ioctl(1, TIOCGWINSZ, &win) != -1 && win.ws_col > 0)
	cmdspc = win.ws_col;
    if (n > 100) n = 100;	/* max of 100 'w' options */
    if (cmdspc > h)
	cmdspc = cmdspc*(n+1) - h + c;
    else
	cmdspc = cmdspc*n + c;
}

/*****************************/
static const char *ttyc(int tty, int pid){
  static char outbuf[8];
  if(CL_num_outp)
    snprintf(outbuf, sizeof outbuf, "%4.4x", tty);
  else
    dev_to_tty(outbuf, 4, tty, pid, ABBREV_TTY|ABBREV_DEV|ABBREV_PTS);
  return outbuf;
}

/*****************************/
/* This is the main driver routine that iterates over the process table.
 */
static void show_procs(unsigned maxcmd, int do_header, int pflags, void* args, int N) {
    static proc_t buf; /* less dynamic memory allocation when not sorting */
    PROCTAB* tab;
    proc_t **ptable = NULL, *next, *retbuf = NULL;
    int n = 0;

    /* initiate process table scan */
    tab = openproc(pflags, args, N);
    if (!tab) {
        fprintf(stderr, "Error: can not access /proc.\n");
        exit(1);
    }

    if (do_header) puts(mode[CL_fmt].header);	/* print header */

    if (!(CL_sort || CL_forest))	/* when sorting and forest are both */
	retbuf = &buf;			/* off we can use a static buffer */

    while ((next = readproc(tab,retbuf))) {	/* read next process */
	n++;					/* Now either: */
	if (CL_forest) {			/*    add process to tree */
	    /* FIXME: this static is disgusting, but to keep binary
	     * compatibility within the 1.2 series I'll just go from
	     * 256 bytes to 1024 bytes, which should be good enough
	     * for most people....  :-(  In the future, the format
	     * format function should probably dynamically allocate
	     * strings.
	     */
	    static char s[1024];
	    (mode[CL_fmt].format)(s, next);
	    if (CL_fmt != PS_V && CL_fmt != PS_M)
		show_time(s+strlen(s), next);
	    add_node(s, next);
	} else if (CL_sort) {			/*    add process to table */
	    ptable = xrealloc(ptable, n*sizeof(proc_t*));
	    ptable[n-1] = next;
	} else {				/*    or show it right away */
	    show_a_proc(&buf, maxcmd);
	    if (buf.cmdline) free((void*)(buf.cmdline[0]));
	    if (buf.environ) free((void*)(buf.environ[0]));
	}
    }
    if (!n) {
	fprintf(stderr, "No processes available.\n");
	exit(1);
    }
    if (CL_sort && !CL_forest) {	/* just print sorted table */
	int i;
	qsort(ptable, n, sizeof(proc_t*), (void*)mult_lvl_cmp);
	for (i = 0; i < n; i++) {
	    show_a_proc(ptable[i], maxcmd);
	    freeproc(ptable[i]);
	}
	free(ptable);
    } else if (CL_forest)
	show_forest();
}

/*****************************/
/* show the trailing command and environment in available space.
 * use abbreviated cmd if requested, NULL list, or singleton NULL string
 */
static void show_cmd_env(char* tskcmd, char** cmd, char** env, unsigned maxch) {
    if (CL_kern_comm)		/* no () when explicit request for tsk cmd */
	maxch = print_str(stdout, tskcmd, maxch);
    else if (!cmd || !*cmd || (!cmd[1] && !*cmd)) {
	/* no /proc//cmdline ==> bounding () */
	if (maxch) {
	    fputc('(', stdout);
	    maxch--;
	}
	maxch = print_str(stdout, tskcmd, maxch);
	if (maxch) {
	    fputc(')', stdout);
	    maxch--;
	}
    } else
	maxch = print_strlist(stdout, cmd, " ", maxch);
    if (CL_show_env && env)
	print_strlist(stdout, env, " ", maxch);
    fputc('\n', stdout);
}


/*****************************/
/* format a single process for output.
 */
static void show_a_proc(proc_t* p, unsigned maxch) {
    static char s[2048];
    (mode[CL_fmt].format)(s, p);
    if (CL_fmt != PS_V && CL_fmt != PS_M)
	show_time(s+strlen(s), p);
    printf("%s", s);
    show_cmd_env(p->cmd, p->cmdline, p->environ, maxch);
}

/****** The format functions for the various formatting modes follow *****/

/*****************************/
static void show_short(char *s, proc_t *p) {
    sprintf(s, "%5d %3s %s", p->pid, ttyc(p->tty,p->pid), status(p));
}


/*****************************/
static void show_long(char *s, proc_t *p) {
    char wchanb[32];
    
    if (GL_wchan_nout)
	sprintf(wchanb, " %-9lx ", p->wchan & 0xffffffff);
    else
	sprintf(wchanb, "%-11.11s", wchan(p->wchan));
    sprintf(s, "%6lx %5d %5d %5d %3ld %3ld %6ld %5ld %-11.11s %s%3s",
	    p->flags, p->euid, p->pid, p->ppid, p->priority, p->nice,
	    p->vsize >> 10, p->rss << (PAGE_SHIFT - 10), wchanb, status(p),
	    ttyc(p->tty,p->pid));
}


/*****************************/
static void show_jobs(char *s, proc_t *p) {
    sprintf(s, "%5d %5d %5d %5d %3s %5d  %s %5d ",
	    p->ppid, p->pid, p->pgrp, p->session, ttyc(p->tty,p->pid), p->tpgid, status(p),
	    p->euid);
}


/*****************************/
static void show_user(char *s, proc_t *p) {
    long pmem, total_time;
    long long seconds;
    time_t start;
    unsigned int pcpu;

    if (CL_num_outp)
	s += sprintf(s, "%5d    ", p->euid);
    else
	s += sprintf(s, "%-8s ", p->euser);
    /* Hertz, being UL, forces 64-bit calculation on 64-bit machines */
    seconds = (((GL_current_time * Hertz) - p->start_time) / Hertz);
    start = GL_time_now - seconds;
    total_time = (p->utime + p->stime +
		  (CL_Sum ? p->cutime + p->cstime : 0));
    pcpu = seconds ?
	(total_time * 10 * 100 / Hertz) / seconds :
	0;
    if (pcpu > 999) pcpu = 999;
    pmem = p->rss * 1000 / (kb_main_total >> (PAGE_SHIFT-10));
    sprintf(s, "%5d %2u.%u %2ld.%ld %5ld %5ld %3s %s %.6s ",
	    p->pid,  pcpu / 10, pcpu % 10,  pmem / 10, pmem % 10,
	    p->vsize >> 10, p->rss << (PAGE_SHIFT - 10), ttyc(p->tty,p->pid), status(p),
	    ctime(&start) + (GL_time_now - start > 3600*24 ? 4 : 10));
}


/*****************************/
/* TODO: this is broken... but does anyone care? */
static void show_sig(char *s, proc_t *p) {
#undef SIGNAL_STRING /* not filled in by the function oldps calls */
#ifdef SIGNAL_STRING
#define TAIL(s) ((s)+(long)strlen(s)-8)
    sprintf(s, "%5d %5d %s %s %s %s %s  %3s ",
	p->euid, p->pid,
	TAIL(p->signal), TAIL(p->blocked), TAIL(p->sigignore), TAIL(p->sigcatch),
	status(p), ttyc(p->tty,p->pid)
    );
#else
    sprintf(s, "%5d %5d %08x %08x %08x %08x %s  %3s ",
	p->euid, p->pid,
	(unsigned)p->signal, (unsigned)p->blocked, (unsigned)p->sigignore, (unsigned)p->sigcatch,
	status(p), ttyc(p->tty,p->pid)
    );
#endif
}


/*****************************/
static void show_vm(char *s, proc_t *p) {
    int pmem;

    s += sprintf(s,"%5d %3s %s", p->pid, ttyc(p->tty,p->pid), status(p));
    show_time(s, p);
    s += strlen(s);
    s += sprintf(s, " %6ld %4ld %4ld %4ld ",
		 p->maj_flt + (CL_Sum ? p->cmaj_flt : 0),
		 p->vsize ? (p->end_code - p->start_code) >> 10 : 0,
		 p->vsize ? (p->vsize - p->end_code + p->start_code) >> 10 : 0,
		 p->rss << (PAGE_SHIFT - 10));
    if(p->rss_rlim == RLIM_INFINITY)
	s += sprintf(s, "   xx ");
    else
	s += sprintf(s, "%5ld ", p->rss_rlim >> 10);
    pmem = p->rss * 1000 / (kb_main_total >> (PAGE_SHIFT-10));
    sprintf(s, "%2d.%d ", pmem / 10, pmem % 10);
}



/*****************************/
static void show_m(char *s, proc_t *p) {
    sprintf(s, "%5d %3s %6ld %6ld %5ld %5ld %5ld %5ld %5ld %5ld %5ld %3ld ", 
	    p->pid, ttyc(p->tty,p->pid),
	    p->maj_flt + (CL_Sum ? p->cmaj_flt : 0),
	    p->min_flt + (CL_Sum ? p->cmin_flt : 0),
	    p->trs << CL_pg_shift,
	    p->drs << CL_pg_shift,
	    p->size << CL_pg_shift,
	    (p->size - p->resident) << CL_pg_shift,
	    p->resident << CL_pg_shift,
	    p->share << CL_pg_shift,
	    p->lrs << CL_pg_shift,
	    p->dt);
}


/*****************************/
static void show_regs(char *s, proc_t *p) {
    char time1[16], time2[16];

    sprintf(s, "%2ld %5d %8lx %8lx %8lx %s %s %s  %3s ",
	    p->start_code >> 26, p->pid, p->start_stack,
	    p->kstk_esp, p->kstk_eip,
	    prtime(time1, p->timeout, GL_current_time*Hertz),
	    prtime(time2, p->it_real_value, 0),
	    status(p), ttyc(p->tty,p->pid));
}


/*****************************/
static char *prtime(char *s, unsigned long t, unsigned long rel) {
    if (t == 0) {
        sprintf(s, "     ");
        return s;
    }
    if ((long) t == -1) {
        sprintf(s, "   xx");
        return s;
    }
    if ((long) (t -= rel) < 0)
        t = 0;
    if (t > 9999)
        sprintf(s, "%5lu", t / 100);
    else
        sprintf(s, "%2lu.%02lu", t / 100, t % 100);
    return s;
}


/*****************************/
static void show_time(char *s, proc_t * p) {
    unsigned t;
    t = (p->utime + p->stime) / Hertz;
    if (CL_Sum) t += (p->cutime + p->cstime) / Hertz;
    sprintf(s, "%3d:%02d ", t / 60, t % 60);
}



/*****************************/
/* fancy process family tree based cmdline printing.  Building the tree
   should be relegated to libproc and only the printing logic should
   remain here.
*/
static struct tree_node * node;  /* forest mode globals */
static int      nodes = 0;
static int      maxnodes = 0;

/*****************************/

static void add_node(char *s, proc_t *task) {
    if (maxnodes == 0) {
	maxnodes = 64;
        node = (struct tree_node *)
            xmalloc(sizeof(struct tree_node) * maxnodes);
    }
    if (nodes >= maxnodes) {
	maxnodes *= 2;
        node = (struct tree_node *)
            xrealloc(node, sizeof(struct tree_node) * maxnodes);
    }
    node[nodes].proc        = task;
    node[nodes].pid         = task->pid;
    node[nodes].ppid        = task->ppid;
    node[nodes].line        = strdup(s);
    node[nodes].cmd         = task->cmd;
    node[nodes].cmdline     = task->cmdline;
    node[nodes].environ     = task->environ;
    node[nodes].children    = 0;
    node[nodes].have_parent = 0;
    nodes++;
}


/*****************************/
static int node_cmp(const void *s1, const void *s2) {
    struct tree_node *n1 = (struct tree_node *) s1;
    struct tree_node *n2 = (struct tree_node *) s2;
    return n1->pid - n2->pid;
}


/*****************************/
static void show_tree(int n, int depth, char *continued) {
    int i, cols = 0;

    fprintf(stdout, "%s", node[n].line);
    for (i = 0; i < depth; i++) {
        if (cols + 4 >= cmdspc - 1)
            break; 
        if (i == depth - 1)
            printf(" \\_ ");
        else if (continued[i])
            printf(" |  ");
        else
            printf("    ");
        cols += 4;
    }
    show_cmd_env(node[n].cmd, node[n].cmdline, node[n].environ, cmdspc - cols);
    for (i = 0; i < node[n].children; i++) {
        continued[depth] = i != node[n].children - 1;
        show_tree(node[n].child[i], depth + 1, continued);
    }
}


/*****************************/
static void show_forest() {
    register int i, j;
    int parent;
    char continued[1024];

    if (CL_sort)
	qsort((void*)node, nodes, sizeof(struct tree_node), (void*)node_mult_lvl_cmp);

    for (i = 0; i < nodes; i++) {
        if (node[i].ppid > 1 && node[i].pid != node[i].ppid) {
	    parent = -1;
	    for (j=0; j<nodes; j++)
		if (node[j].pid==node[i].ppid)
		    parent = j;
        } else
            parent = -1;
        if (parent >= 0) {
            node[i].have_parent++;
            if (node[parent].children == 0) {
                node[parent].child = (int*)xmalloc(16 * sizeof(int*));
                node[parent].maxchildren = 16;
            }
            else if (node[parent].children == node[parent].maxchildren) {
                node[parent].maxchildren *= 2;
                node[parent].child = (int*)xrealloc(node[parent].child,
						    node[parent].maxchildren
						    * sizeof(int*));
            }
            node[parent].child[node[parent].children++] = i;
        }
    }

    for (i = 0; i < nodes; i++) {
        if (!node[i].have_parent)
            show_tree(i, 0, continued);
    }
}
